service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/auth/TokenCredentialCache.java

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.service.auth;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+
+/**
+ * This class generates AAD authentication tokens from a TokenCredential but caches previous tokens when they aren't near
+ * expiry.
+ */
+public class TokenCredentialCache
+{
+    private final static int MINUTES_BEFORE_PROACTIVE_RENEWAL = 9;
+    private final TokenCredential tokenCredential;
+    private AccessToken accessToken;
+
+    /**
+     * Construct a new TokenCredentialCache instance.
+     * @param tokenCredential The tokenCredential instance that this cache will use to generate new tokens.
+     */
+    public TokenCredentialCache(TokenCredential tokenCredential)
+    {
+        Objects.requireNonNull(tokenCredential, "tokenCredential cannot be null");
+
+        this.tokenCredential = tokenCredential;
+    }
+
+    /**
+     * Get a valid AAD authentication token. This may be the same as a previously returned token if it is not near expiration time yet.
```

Comment:
Would be good to add an explanation of what near expiration time is in the implementation here.